### PR TITLE
tests: linkcheck: add client connection-pool contention coverage

### LINKED ISSUE #11324: linkcheck builder: begin using requests.Session functionality during linkchecking
**Is your feature request related to a problem? Please describe.**
At the moment, the `linkcheck` builder performs individual `request.get` (or similar HTTP request method) operations during linkchecking, without any explicit connection or session pooling.

This may be inefficient, because it seems likely that for many use cases, linkchecking will make multiple requests to the same host (because documentation references are likely to have host-locality).

**Describe the solution you'd like**
Confirmation that connection pooling is not currently in use would be a good starting point; in other words: we should confirm that linkchecking of multiple URLs on a single host results in multiple TCP connections.  Ideally this should be written as a test case.

If we can confirm that the problem exists, then we may be able to use some of the [`Session` object functionality](https://requests.readthedocs.io/en/latest/user/advanced/#session-objects) from the `requests` library that's already in use here to enable connection pooling.

**Describe alternatives you've considered**
None so far, although open to suggestions (and improvements on the definition of this feature request).

**Additional context**
- See note / suggestion about use of context managers: https://github.com/sphinx-doc/sphinx/issues/11317#issuecomment-1508319197

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Test coverage

### Purpose
- Provides test coverage for implementations attempting to resolve #11324.

### Detail
- Configure an artificially-limited connection pool by patching `requests.Session.get_adapter`
- Create a multi-item `linkcheck` workload
- Consume one item at a time from the workload from separate threads
- Ensure that all items were consumed and that no timeouts occurred

### Relates
- Prepares for #11324.
- Planned from https://github.com/sphinx-doc/sphinx/issues/11348#issuecomment-1532656258.

### CODE REVIEW

**Comment by jayaddison** [comment]
Why does the `test_connection_contention` test case on Windows only manage to perform [4 of the 10 linkchecks](https://github.com/sphinx-doc/sphinx/actions/runs/4915585033/jobs/8778252933?pr=11402#step:5:1966) before the [5-second timeout](https://github.com/jayaddison/sphinx/blob/60765d3e655d11ac3a9b4eabc8723c4ea8c96c7e/tests/test_build_linkcheck.py#L687) occurs?

File: tests/test_build_linkcheck.py [hunk 1]
```diff
@@ -661,6 +666,45 @@ def test_limit_rate_bails_out_after_waiting_max_time(app):
...
 
+@mock.patch('sphinx.util.requests.requests.Session.get_adapter')
+def test_connection_contention(get_adapter, app, capsys):
+    # Create a shared, but limited-size, connection pool
+    import requests
+    get_adapter.return_value = requests.adapters.HTTPAdapter(pool_maxsize=1)
+
+    # Set an upper-bound on socket timeouts globally
+    import socket
+    socket.setdefaulttimeout(5)
```
Comment (jayaddison) [inline tests/test_build_linkcheck.py:780]: I'd like to remove this because I think it has side-effects that outlast the unit test.

It was introduced to avoid indefinitely-blocked socket I/O when the test linkcheck client and server fail to communicate reliably.

**Comment by jayaddison** [comment]
> Why does the `test_connection_contention` test case on Windows only manage to perform [4 of the 10 linkchecks](https://github.com/sphinx-doc/sphinx/actions/runs/4915585033/jobs/8778252933?pr=11402#step:5:1966) before the [5-second timeout](https://github.com/jayaddison/sphinx/blob/60765d3e655d11ac3a9b4eabc8723c4ea8c96c7e/tests/test_build_linkcheck.py#L687) occurs?

I've got a strange theory about this now.  Roughly speaking, I expect the problem to be either something to do with sockets, threading, or DNS (all areas that could have significantly-different behaviour per-platform).

The theory I have is to do with an existing Python bug regarding sockets on Windows _when timeouts are enabled_.  I won't link to the bug yet before confirming whether it's relevant, so as not to pollute it with noise, but it is bug number eight five nine seven two, for reference.

Basically it seems that when timeouts are enabled, Windows likes to wait until a minimum number of bytes are in the socket network buffer(s) before sending the data.  That _could_ explain why either the client or server aren't communicating completely.

I think that loading more test links into the worker queue could be one way to confirm this - basically to front-load enough traffic to ensure that the client and server complete the communication required for the unit test.

**Comment by jayaddison** [comment]
Nope!  So much for that theory.  Exactly the same failure: only 4 out of 10 links checked when running the `test_connection_contention` test on Windows.

**Comment by jayaddison** [comment]
Ok, well that worked.. but maybe the excess workload wasn't necessary for the fix (the parallelism may have allowed for enough items to be consumed in parallel).  So: a small cleanup next to remove the excess linkcheck requests.

**Comment by jayaddison** [comment]
> I'll leave https://github.com/sphinx-doc/sphinx/commit/589bca53c50d8dee080bf18b48ae42231e8676b3 in place and refactor the test so that the threads are initiated in parallel (although they'll still only have a single connection available in the `urllib3` pool to use).

The potential problem with that refactor is that it means that each thread is no longer guaranteed to consume a single item -- nor guaranteed (in theory) to contend with any other threads.  In practice connection contention probably will occur.. I'm trying to think of more reliable approaches.

File: tests/test_build_linkcheck.py (hunk 1 above)
Comment (francoisfreitag) [reply tests/test_build_linkcheck.py:780]: I believe this could be a global setting for the test suite, I don’t see why an indefinitely-blocked socket I/O would be reasonable in that context.

Comment (jayaddison) [reply tests/test_build_linkcheck.py:780]: Yep, that would make sense.  It would also be nice if the tests could fail if that timeout limit is ever reached, because that'd almost certainly be a sign of a problem with the application code, given that we now entirely request local, non-Internet resources during tests (I believe that's true?).

Another relevant note is that, in practice, I think we already approximate this behaviour -- perhaps unintentionally -- from the time at which a linkchecker object is first constructed.  That's because there is a call to the same `setdefaulttimeout` method here:

https://github.com/sphinx-doc/sphinx/blob/d48cc783b6e48b99cbbcf5462f44c59a4ed78e6a/sphinx/builders/linkcheck.py#L119-L120

It does _probably_ make sense to have a timeout in production environments, because in most cases it would be bad for a single long-delayed link to block the entire linkcheck process, but at the same time it would be good to confirm why the global timeout is in place given that there is an existing -- and configurable -- linkcheck timeout setting:

https://github.com/sphinx-doc/sphinx/blob/d48cc783b6e48b99cbbcf5462f44c59a4ed78e6a/sphinx/builders/linkcheck.py#L270-L271

(I also should explain that it seems OK to me that _both_ timeouts exist: I sometimes have a tendency to remove redundancy in software to such an extent that the resulting code becomes fragile in other ways, and I'm keen that we avoid that risk here.  but I think it would be good to understand the reasoning why both exist.  I think it's basically to try to avoid the linkchecker ever stalling indefinitely, even in the absence of a configured `linkcheck_timeout` value)

Comment (francoisfreitag) [reply tests/test_build_linkcheck.py:780]: I agree with you. My reading of the code (and https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_timeout) is:
1. `linkcheck_timeout=None`: make sure linkcheck bails out after a reasonable amout of time (5.0 seconds), since `requests` won’t handle the timeout
2. `linkcheck_timeout=3.14`: specify the timeout to `requests`

So, by default, `linkcheck` will not get stuck. When the default does not fit a project (developers want slower or longer timeouts), they can use `linkcheck_timeout`. I don’t see a way to specify no timeout, but I don’t think that’s an issue. A noteworthy surprise is that setting `linkcheck_timeout=5.0` is different from using the default, as the default only matters for the socket connection, whereas [`requests` timeout also considers the time to generate the response on the server](https://docs.python-requests.org/en/latest/user/advanced/#timeouts).

I think the side-effect is not intentional but is pretty nice to have and should not be changed. There aren’t other `socket.setdefaulttimeout()` in the app, so linkcheck sets that default for other extensions (thinking of extlinks). My suggestion is to move the `socket.setdefaulttimeout()` to `sphinx.application.Sphinx.__init__()`. That makes the call explicit and should not break BC (since the `socket.setdefaulttimeout()` happened when the extensions were being setup, as `setup_extension` imports the module).

I believe the test suite could use the same setting as the application, or even a shorter one, considering that waiting 5 seconds is pretty long in that context. Anything above a second would probably indicate an issue IMO.


Coming back to `linkcheck`, we could remove the call to `socket.setdefaulttimeout()` (that has a side effect on the Sphinx application), and set a default of `5.0` for `linkcheck_timeout`, but it would break BC:
- https://docs.python-requests.org/en/latest/user/advanced/#timeouts shows `socket.setdefaulttimeout()` will timeout only on what `requests` calls “connect”, not while reading data from the server. So a user who connected quickly but waited a long time to receive the response would see failures where it previously worked. I’m guessing we could maybe (and I don’t wanna) abuse the current implementation to pass a tuple `(5.0, None)` and preserve the current behavior.
- users may call `socket.setdefaulttimeout()` after `Sphinx.__init__()`, to use their own values. That would be a bit surprising considering `linkcheck_timeout` exists, but who knows...

Comment (francoisfreitag) [reply tests/test_build_linkcheck.py:780]: > That makes the call explicit and should not break BC (since the socket.setdefaulttimeout() happened when the extensions were being setup, as setup_extension imports the module).

That’s actually not true. `socket.setdefaulttimeout(5.0)` is only called during `CheckExternalLinksBuilder.__init__()`, which happens when the builder is created, not when the module is imported and the builder registered.

Comment (jayaddison) [reply tests/test_build_linkcheck.py:780]: > I agree with you. My reading of the code (and https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_timeout) is:
> 
>     1. `linkcheck_timeout=None`: make sure linkcheck bails out after a reasonable amout of time (5.0 seconds), since `requests` won’t handle the timeout

I'm not so sure about this - my reading of the documentation:

> linkcheck_timeout
>
>    A timeout value, in seconds, for the linkcheck builder. The default is to use Python’s global socket timeout.
>
>    New in version 1.1.

... is that the unmodified global timeout for sockets is to be expected unless the operator changes the configuration value.

What we have currently is a 5-second timeout set by the application when no configuration is applied.  I do think that having a timeout in production makes sense, but either the documentation or the code may need updating to bring them into consistency.

It'll take me a while longer to process the rest of your reply and to think about any further changes.

**Comment by jayaddison** [comment]
Thanks @AA-Turner!

File: tests/test_build_linkcheck.py (hunk 1 above)
Comment (jayaddison) [reply tests/test_build_linkcheck.py:780]: I'll create a separate issue soon to track resolution for this (relocation and/or removal of the `socket.setdefaulttimeout` calls).
